### DISCUSSION TITLE
[Proposal] Throw exceptions for DB errors

### DISCUSSION THREAD

**Discussion by prettyboymp** [woocommerce/woocommerce#43275] [General]
## Proposal

Any place where WooCommerce core (or extensions) are directly querying the database for commerce data, should throw an exception if that query fails.

## Introduction

The default handling for query exceptions within WordPress is to essentially ignore them and carry on.  For the most part, this isn't a big deal for blog/website content.  However, for commerce and financial data, this is a much bigger concern.  The  failed DB queries can cause data integrity issues that go unnoticed until they cause failures later on without a great context as to why.

The idea of throwing an exception on DB failures is being proposed as an initial step of improving overall fault integrity of the WooCommerce Platform.  This first step will allow us to stop execution of the current process before more damage can be done by the failed query or invalid empty result.  Once we are throwing these exceptions, we can then look to catch and handle them properly based on the higher level context; e.g. retrying idempotent actions.

## Background

We've recently had a couple of incidents with extremely high database load caused by the combination of HPOS/Post table data syncing and webhooks.  We've noticed that some orders and subscriptions that were created and/or modified during this time have some data inconsistencies that we believe are tied to MySQL queries that timed out or failed during these events.

Some of the data integrity issues involved:
* Incorrectly updated date_paid or date_completed dates on old orders (believed to be triggered during subscription renewal events)
* Duplicate subscriptions being created 
* Missing payment meta

## Possible Implementation

Introduce a function similar to the below to wrap the results of a `$wpdb->query()` (or equivalent) call.

```
function throw_on_wpdb_error( $pass_through_result = null ) {
	global $wpdb;

	if ( $wpdb->last_error ) {
		throw new QueryException( $wpdb->last_error, $wpdb->last_errno, $wpdb->last_query );
	}

	return $pass_through_result;
}
```

**Comment by mattsherman** [comment]
I like the idea of throwing exception for this, since then errors can't be silently ignored.

**Comment by vedanshujain** [comment]
Hey @prettyboymp we discussed this today. I think it would be a good idea for WooCommerce to somehow support more strictness around its database errors instead of silently ignoring them, for the reasons that you mention, but what's not immediately clear is how we will go about doing that.

If we add a blanket query wrapper like you suggested, then it would affect everything and not just WC query parts. Historically, when we have tried to add strictness like this, it has bounced back since it would be a significant change. There's also the question, if we are willing to make an invasive change like this, should it be WooCommerce territory or something that should be addressed in WP Core's wpdb class?

We are also hoping to look at it in different direction, for example in [#43770](https://github.com/woocommerce/woocommerce/issues/43770) which is not exactly what you are suggesting, but is aiming at the same root cause. Maybe that's sufficient enough?

wdyt?
